File: templates/python/hot-100/p7_linked_list/t138_copy_random_list.py
"""LeetCode 138. Copy List with Random Pointer
https://leetcode.cn/problems/copy-list-with-random-pointer/
"""

from __future__ import annotations

from typing import Optional


class Node:
    def __init__(
        self,
        val: int = 0,
        next: Optional[Node] = None,
        random: Optional[Node] = None,
    ):
        self.val = val
        self.next = next
        self.random = random


class Solution:
    def copyRandomList(self, head: Optional[Node]) -> Optional[Node]:
        pass

File: templates/java/hot-100/p7_linked_list/T138_copyRandomList.java
/**
 * Practice template for:
 * - 138. Copy List with Random Pointer
 *   https://leetcode.cn/problems/copy-list-with-random-pointer/
 */
import java.util.HashMap;
import java.util.Map;

/**
 * @author Cheng Yihao
 * @version 1.0
 * @date 2025/11/9 17:41
 * @comment
 */
public class T138_copyRandomList {


    public Node copyRandomList(Node head) {
        if (head == null) {
            throw new UnsupportedOperationException("TODO");
        }

        Node curr = head;
        Map<Node, Node> map = new HashMap<>();

        while (curr != null) {
            Node copyNode = new Node(curr.val);
            map.put(curr, copyNode);
            curr = curr.next;
        }

        curr = head;
        while (curr != null) {
            Node copyNode = map.get(curr);
            copyNode.next = map.get(curr.next);
            copyNode.random = map.get(curr.random);
            curr = curr.next;
        }
        return map.get(head);
    }

    class Node {
        int val;
        Node next;
        Node random;

        public Node(int val) {
            this.val = val;
            this.next = null;
            this.random = null;
        }
    }

}
